File: tests/fixtures/go_repo/internal/services/user_service.go
// Package services — user account business logic.
package services

import (
	"context"
	"crypto/rand"
	"encoding/hex"
	"errors"
	"fmt"

	"github.com/example/goapi/internal/models"
	"github.com/example/goapi/internal/repository"
	"github.com/example/goapi/internal/types"
)

// ErrEmailAlreadyTaken is returned by CreateUser when the supplied email is
// already registered.
var ErrEmailAlreadyTaken = errors.New("email already taken")

// UserService orchestrates user account creation, lookup, and updates.
type UserService struct {
	repo *repository.UserRepository
	auth *AuthService
}

// NewUserService constructs a UserService wired to its dependencies.
func NewUserService(repo *repository.UserRepository, auth *AuthService) *UserService {
	return &UserService{repo: repo, auth: auth}
}

// CreateUser registers a new account. The password is bcrypt-hashed before
// being persisted. Returns ErrEmailAlreadyTaken on email collision.
func (s *UserService) CreateUser(ctx context.Context, req *types.CreateUserRequest) (*models.User, error) {
	if existing, _ := s.repo.FindByEmail(ctx, req.Email); existing != nil {
		return nil, ErrEmailAlreadyTaken
	}
	hash, err := s.auth.HashPassword(req.Password)
	if err != nil {
		return nil, fmt.Errorf("hash password: %w", err)
	}
	user := models.NewUser(newID(), req.Email, req.Username, hash)
	if err := s.repo.Insert(ctx, user); err != nil {
		return nil, fmt.Errorf("insert user: %w", err)
	}
	return user, nil
}

// GetUserByID fetches a user by their primary key.
func (s *UserService) GetUserByID(ctx context.Context, id string) (*models.User, error) {
	return s.repo.FindByID(ctx, id)
}

// GetUserByEmail fetches a user by their email address.
func (s *UserService) GetUserByEmail(ctx context.Context, email string) (*models.User, error) {
	return s.repo.FindByEmail(ctx, email)
}

// ListUsers returns a paginated slice of users.
func (s *UserService) ListUsers(ctx context.Context, page, pageSize int) ([]models.User, int, error) {
	if page < 1 {
		page = 1
	}
	if pageSize < 1 || pageSize > 100 {
		pageSize = 20
	}
	offset := (page - 1) * pageSize
	rows, err := s.repo.List(ctx, offset, pageSize)
	if err != nil {
		return nil, 0, err
	}
	total, err := s.repo.Count(ctx)
	if err != nil {
		return nil, 0, err
	}
	return rows, total, nil
}

// UpdateUser applies a patch to an existing user record.
func (s *UserService) UpdateUser(ctx context.Context, id string, patch *types.UpdateUserRequest) (*models.User, error) {
	current, err := s.repo.FindByID(ctx, id)
	if err != nil {
		return nil, err
	}
	if patch.Email != nil {
		current.Email = *patch.Email
	}
	if patch.Username != nil {
		current.Username = *patch.Username
	}
	if patch.Password != nil {
		hash, err := s.auth.HashPassword(*patch.Password)
		if err != nil {
			return nil, fmt.Errorf("hash password: %w", err)
		}
		current.PasswordHash = hash
	}
	if err := s.repo.Update(ctx, id, current); err != nil {
		return nil, err
	}
	return current, nil
}

// DeleteUser removes a user by id.
func (s *UserService) DeleteUser(ctx context.Context, id string) error {
	return s.repo.Delete(ctx, id)
}

// newID returns a random 128-bit hex identifier suitable for user/item IDs.
func newID() string {
	var b [16]byte
	_, _ = rand.Read(b[:])
	return hex.EncodeToString(b[:])
}

File: tests/fixtures/rust_repo/src/services/user_service.rs
//! User account business logic — signup, lookup, update, delete.

use std::sync::Arc;

use uuid::Uuid;

use crate::dto::user_dto::{CreateUserRequest, UpdateUserRequest};
use crate::error::ApiError;
use crate::models::User;
use crate::repository::user_repository::UserRepository;
use crate::services::auth_service::AuthService;

/// Orchestrates the user lifecycle. Owns hashing-then-persisting in
/// [`UserService::create_user`] so handlers can't accidentally skip the hash
/// step.
pub struct UserService {
    repo: Arc<UserRepository>,
    auth: Arc<AuthService>,
}

impl UserService {
    /// Construct a `UserService` wired to its dependencies.
    pub fn new(repo: Arc<UserRepository>, auth: Arc<AuthService>) -> Self {
        Self { repo, auth }
    }

    /// Register a new user account. The password is argon2-hashed before
    /// hitting the database. Returns [`ApiError::Conflict`] on email
    /// collision.
    pub async fn create_user(&self, req: &CreateUserRequest) -> Result<User, ApiError> {
        if self.repo.find_by_email(&req.email).await.is_ok() {
            return Err(ApiError::Conflict("email already taken".into()));
        }
        let hash = self.auth.hash_password(&req.password)?;
        let user = User::new(Uuid::new_v4(), req.email.clone(), req.username.clone(), hash);
        self.repo.insert(&user).await?;
        Ok(user)
    }

    /// Fetch a user by primary key.
    pub async fn get_user_by_id(&self, id: Uuid) -> Result<User, ApiError> {
        Ok(self.repo.find_by_id(id).await?)
    }

    /// Fetch a user by email address.
    pub async fn get_user_by_email(&self, email: &str) -> Result<User, ApiError> {
        Ok(self.repo.find_by_email(email).await?)
    }

    /// Return a paginated slice of users along with the total count for
    /// page-meta. Clamps `page` to `>= 1` and `page_size` to `[1, 100]`.
    pub async fn list_users(
        &self,
        page: u32,
        page_size: u32,
    ) -> Result<(Vec<User>, i64), ApiError> {
        let page = page.max(1);
        let page_size = page_size.clamp(1, 100);
        let offset = ((page - 1) * page_size) as i64;
        let rows = self.repo.list(offset, page_size as i64).await?;
        let total = self.repo.count().await?;
        Ok((rows, total))
    }

    /// Apply a patch to an existing user record. Any `None` field on the
    /// request is left untouched.
    pub async fn update_user(
        &self,
        id: Uuid,
        patch: &UpdateUserRequest,
    ) -> Result<User, ApiError> {
        let mut current = self.repo.find_by_id(id).await?;
        if let Some(email) = patch.email.as_ref() {
            current.email = email.clone();
        }
        if let Some(username) = patch.username.as_ref() {
            current.username = username.clone();
        }
        if let Some(password) = patch.password.as_ref() {
            current.password_hash = self.auth.hash_password(password)?;
        }
        self.repo.update(&current).await?;
        Ok(current)
    }

    /// Remove a user by id.
    pub async fn delete_user(&self, id: Uuid) -> Result<(), ApiError> {
        Ok(self.repo.delete(id).await?)
    }
}
